rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+use crate::barriers::{init_with_barrier, write_with_barrier};
+use crate::memory::{alloc_blob, Memory};
+use crate::rts_trap_with;
+use crate::types::{size_of, Blob, Bytes, Region, Value, TAG_REGION};
+
+use motoko_rts_macros::ic_mem_fn;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct BlockId(pub u16);
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct RegionId(pub u16);
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct RegionSizeInPages(pub u64);
+
+// Note: Use this type only in local variables, as it contains raw pointers
+// that are not handled by the GCs.
+#[derive(Clone)]
+pub struct AccessVector(pub *mut Blob);
+
+// Note: Use this type only in local variables, as it contains raw pointers
+// that are not handled by the GCs.
+#[derive(Clone)]
+pub struct RegionObject(pub *mut Region);
+
+const NIL_REGION_ID: u16 = 0;
+
+// For giving the other logic the right number at the right time.
```

Comment:
Maybe, some brief comments on the purpose of these three variables `REGION_SET_MEM_SIZE`, `REGION_MEM_SIZE_INIT`, `REGION_TOTAL_ALLOCATED_BLOCKS` and their invariants could be helpful.
If I understand correctly,  `REGION_SET_MEM_SIZE` is only used in two cases and has higher precedence than `REGION_TOTAL_ALLOCATED_BLOCKS`:
(1) With stable region support: After destabilization until the first stable region growth (which may also happen during the migration logic).
(2) If stable region is disabled (legacy modes).
Either `REGION_TOTAL_ALLOCATED_BLOCKS` or `REGION_SET_MEM_SIZE` is set (to a non-default value), but not both.

Reply:
#4143 manages to remove REGION_SET_MEM_SIZE and region_set/get_mem_size so hopefully is easier to grok.